New Span-based C# parsing logic

jtattermusch opened the pull request:
A big refactor of C# parsing logic to allow parsing from ReadOnlySequence<byte> (which would allow parsing directly from the data as received on the wire without requiring to copy the buffers and thus reducing the GC pressure a lot).

Main changes:
- refactor out the parsing logic to helper classes (to avoid any duplication of the parsing logic)
- store all the parsing state into a struct (currently called ParserInternalState)
- introduce a ParseContext ref struct that has public methods, but they are only meant to be invoked by the generated code
- generate a InternalMergeFrom(ref ParseContext ctx) method, which has the actual parsing code (and is only supposed to be called by the generated code, never directly by the users.
- allow easy and fast conversion from CodedInputStream into a ParseContext (because the parsing state is encapsulated in the ParserInternalState, so bringing the state between the two only requires copying a struct.
- the original IMessage.MergeFrom(CodedInputStream cis) methods redirects to InternalMergeFrom(ref ParseContext). This transition has some (very small = copying a struct) overhead, but is basically only done once in the entire parsing tree, so that seem acceptable.
- all the public APIs that take CodedInputStream still behave exactly the same.
- compatibility with old generated code: if one uses old generated code against a newer runtime, everything will still work, but there might be some performance penalty (see benchmarks). Regenerating the code will speed things up.
- the public API increase is kept as small as possible. Basically the only new method we provide is Parse.ParseFrom(ReadOnlySequence<byte> byte) as entry point for the more efficient parsing. There are a few other additions (parsing maps, repeated fields etc.), but they are only meant to be invoked by the generated code.
- this PR only addresses memory-efficient parsing. Memory efficient serialization can be added in a followup PR.

This PR drops netstandard1.0 support (which doesn't support the Span type we need) and adds netstandard1.1 instead. The impact on supported platforms should be minimal (almost everyone supports netstandard2.0 these days anyway).

Basically this PR a new and simpler take on https://github.com/protocolbuffers/protobuf/pull/5888 (which is much bigger, introduces much more code duplication and leads to bigger public API increase).

jtattermusch commented:
CC @JamesNK @JunTaoLuo 

this is still WIP and I'll probably do some rebasing, but all this seems to be the minimal PR that adds ReadOnlySequence parsing.

JamesNK commented:
Changes to fix running on .NET Framework:

https://github.com/JamesNK/protobuf/commit/5d868ad9f34c0f850ff0ba450d75370d1cb9a8cf

JamesNK commented on csharp/src/Google.Protobuf/IBufferMessage.cs:
```diff
@@ -0,0 +1,49 @@
+        /// Internal implementation of merging data from given parse context into this message.
+        /// Users should never invoke this method directly.
+        /// </summary>        
+        void MergeFrom_Internal(ref ParseContext ctx);
```
I'm not a big fan of `_Internal` suffix.

I'd rather MergeFromContext, or MergeFromInternal, or even just MergeFrom and this comment warning that this overload of MergeFrom shouldn't be used.

JamesNK replied:
What are your plans for message writing? WriteContext and an equivalent method here?

I wonder if IBufferMessage is still the right name for this interface.

jtattermusch replied:
I think MergeFromInternal and InternalMergeFrom also works.
Explicitly implementing the interface in generated code sounds like a good idea - I'll try that.

For message writing I think we can use a similar approach (which will hopefully require way fewer code changes than parsing). I think serialization could go in as a follow-up PR to make this PR smaller (we can make sure we won't release the intermediate version)

Do you have suggestions for how IBufferMessage should be renamed?

JamesNK commented on csharp/src/Google.Protobuf/MessageParser.cs:
```diff
@@ -113,6 +115,19 @@ public IMessage ParseFrom(Stream input)
+        /// <param name="data">The data to parse.</param>
+        /// <returns>The parsed message.</returns>
+        [SecuritySafeCritical]
+        public IMessage ParseFrom(ReadOnlySequence<byte> data)
```
Only used in benchmarks

jtattermusch replied:
See GeneratedMessageTest improvements.

JamesNK commented on csharp/src/Google.Protobuf/MessageParser.cs:
```diff
@@ -287,6 +302,19 @@ public new T ParseFrom(Stream input)
+        /// <param name="data">The data to parse.</param>
+        /// <returns>The parsed message.</returns>
+        [SecuritySafeCritical]
+        public new T ParseFrom(ReadOnlySequence<byte> data)
```
Only used in benchmarks

jtattermusch replied:
I improved the GeneratedMessageTest so that it verifies multiple codepaths (I borrowed the MessageParsingHelpers from your original PR which works very well for that purpose) and we should have some basic coverage now.

jtattermusch commented on csharp/src/Google.Protobuf/ParseContext.cs:
```diff
@@ -0,0 +1,332 @@
+        internal ParserInternalState state;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Initialize(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, out ParseContext ctx)
```
With the out parameter we can write directly into the destination struct (even if it's a subfield of a bigger struct which is something we utilize when initializing structs nested under ParseContext).
ref return might work as well, but I haven't experimented with it. Given this is purely internal code, I'd say out param works fine for now and we can later experiments with ref struct and compare the performance.

Btw, I was somewhat surprised that setting myStruct.subStruct = new XYZStruct() seems to involve a copy (the compiler could optimize that to initialize subStruct in place). This is basically the reason why I used the out param in the first place.

jtattermusch commented on csharp/src/Google.Protobuf/ParseContext.cs:
```diff
@@ -0,0 +1,332 @@
+        internal ParserInternalState state;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Initialize(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, out ParseContext ctx)
```
most invocations of the parsing primitives take `ref span, ref state`, so I used that also here for consistency.

JamesNK commented:
@googlebot I consent.

jtattermusch commented:
ParseRawPrimitivesBenchmark results:

raw parsing speed is approx the same when you compare the original timing (e.g. ParseRawVarint32) with the _ParseContext variant (e.g. ParseRawVarint32_ParseContext).
For some values this PR is faster than the baseline and for others it's slower, but IMHO the results seem good enough. I also assume that ParseContext-based parsing can be fine-tuned in follow-up pull requests.

comparing the original timing against _CodedInputStream: this is the "backwards compatibility" codepath when someone wants to use CodedInputStream instances directly or when legacy generated code is used. There is some performance hit (we need to convert the internal byte[] buffer into a span with each invocation of CodedInputStream.ReadXYZ()), but because this codepath will be only hit in compatibility scenarios, the results seem totally acceptable.

```
BEFORE

|           Method | BytesToParse | encodedSize |      Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------- |------------- |------------ |----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| ParseRawVarint32 |        10080 |           1 | 42.587 us | 0.8495 us | 1.3957 us |           - |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           1 | 45.554 us | 0.8631 us | 0.8074 us |           - |           - |           - |                88 B |
| ParseRawVarint32 |        10080 |           2 | 28.323 us | 0.1714 us | 0.1431 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           2 | 25.848 us | 0.2385 us | 0.2231 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint32 |        10080 |           3 | 20.259 us | 0.1440 us | 0.1203 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           3 | 24.808 us | 0.2482 us | 0.2073 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint32 |        10080 |           4 | 17.403 us | 0.2847 us | 0.2663 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           4 | 22.338 us | 0.2093 us | 0.1855 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint32 |        10080 |           5 | 15.288 us | 0.1741 us | 0.1543 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           5 | 21.688 us | 0.1247 us | 0.1167 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           6 | 21.317 us | 0.1854 us | 0.1643 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           7 | 20.995 us | 0.4196 us | 0.3925 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           8 | 20.677 us | 0.2127 us | 0.1776 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |           9 | 20.784 us | 0.3234 us | 0.2700 us |      0.0305 |           - |           - |                88 B |
| ParseRawVarint64 |        10080 |          10 | 20.433 us | 0.1480 us | 0.1384 us |      0.0305 |           - |           - |                88 B |
|     ParseFixed32 |        10080 |           ? | 12.097 us | 0.0976 us | 0.0815 us |      0.0305 |           - |           - |                88 B |
|     ParseFixed64 |        10080 |           ? |  6.754 us | 0.1340 us | 0.1316 us |      0.0381 |           - |           - |                88 B |
|    ParseRawFloat |        10080 |           ? | 10.141 us | 0.0834 us | 0.0740 us |      0.0305 |           - |           - |                88 B |
|   ParseRawDouble |        10080 |           ? |  5.818 us | 0.1146 us | 0.1125 us |      0.0381 |           - |           - |                88 B |



AFTER

BenchmarkDotNet=v0.11.4, OS=macOS Mojave 10.14.6 (18G3020) [Darwin 18.7.0]
Intel Core i7-7567U CPU 3.50GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT


|                            Method | BytesToParse | encodedSize |      Mean |     Error |    StdDev |    Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------------- |------------- |------------ |----------:|----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| ParseRawVarint32_CodedInputStream |        10080 |           1 | 60.206 us | 1.3361 us | 1.2498 us | 59.732 us |           - |           - |           - |               184 B |
|     ParseRawVarint32_ParseContext |        10080 |           1 | 40.209 us | 0.1798 us | 0.1501 us | 40.205 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           1 | 50.539 us | 0.7206 us | 0.6741 us | 50.214 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           1 | 46.684 us | 0.2891 us | 0.2414 us | 46.623 us |           - |           - |           - |                   - |
| ParseRawVarint32_CodedInputStream |        10080 |           2 | 34.887 us | 0.3371 us | 0.3153 us | 34.865 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint32_ParseContext |        10080 |           2 | 26.858 us | 0.2962 us | 0.2626 us | 26.729 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           2 | 32.205 us | 0.2986 us | 0.2647 us | 32.122 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           2 | 32.317 us | 0.2829 us | 0.2209 us | 32.293 us |           - |           - |           - |                   - |
| ParseRawVarint32_CodedInputStream |        10080 |           3 | 23.401 us | 0.2077 us | 0.1841 us | 23.337 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint32_ParseContext |        10080 |           3 | 19.718 us | 0.1820 us | 0.1520 us | 19.721 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           3 | 26.057 us | 0.2801 us | 0.2339 us | 26.028 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           3 | 26.161 us | 0.3419 us | 0.3198 us | 26.078 us |           - |           - |           - |                   - |
| ParseRawVarint32_CodedInputStream |        10080 |           4 | 19.313 us | 0.0898 us | 0.0750 us | 19.313 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint32_ParseContext |        10080 |           4 | 18.223 us | 0.1677 us | 0.1486 us | 18.217 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           4 | 22.932 us | 0.2703 us | 0.2396 us | 22.851 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           4 | 22.799 us | 0.1112 us | 0.0929 us | 22.783 us |           - |           - |           - |                   - |
| ParseRawVarint32_CodedInputStream |        10080 |           5 | 18.677 us | 0.4719 us | 0.5435 us | 18.443 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint32_ParseContext |        10080 |           5 | 16.927 us | 0.1420 us | 0.1186 us | 16.889 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           5 | 20.978 us | 0.1443 us | 0.1279 us | 20.922 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           5 | 21.781 us | 0.4325 us | 1.1692 us | 21.304 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           6 | 19.808 us | 0.1435 us | 0.1272 us | 19.804 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           6 | 19.910 us | 0.4876 us | 0.6674 us | 19.663 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           7 | 18.991 us | 0.1813 us | 0.1608 us | 18.952 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           7 | 19.064 us | 0.2787 us | 0.2470 us | 19.018 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           8 | 18.439 us | 0.2083 us | 0.1847 us | 18.384 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           8 | 18.301 us | 0.1620 us | 0.1265 us | 18.266 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |           9 | 17.941 us | 0.1010 us | 0.0944 us | 17.963 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |           9 | 17.685 us | 0.1960 us | 0.1737 us | 17.651 us |           - |           - |           - |                   - |
| ParseRawVarint64_CodedInputStream |        10080 |          10 | 17.653 us | 0.0712 us | 0.0595 us | 17.631 us |      0.0610 |           - |           - |               184 B |
|     ParseRawVarint64_ParseContext |        10080 |          10 | 17.219 us | 0.0882 us | 0.0825 us | 17.179 us |           - |           - |           - |                   - |
|     ParseFixed32_CodedInputStream |        10080 |           ? | 10.826 us | 0.1526 us | 0.1274 us | 10.810 us |      0.0763 |           - |           - |               184 B |
|         ParseFixed32_ParseContext |        10080 |           ? |  7.372 us | 0.0512 us | 0.0454 us |  7.361 us |           - |           - |           - |                   - |
|     ParseFixed64_CodedInputStream |        10080 |           ? |  4.063 us | 0.0378 us | 0.0316 us |  4.054 us |      0.0839 |           - |           - |               184 B |
|         ParseFixed64_ParseContext |        10080 |           ? |  3.409 us | 0.0183 us | 0.0143 us |  3.407 us |           - |           - |           - |                   - |
|    ParseRawFloat_CodedInputStream |        10080 |           ? | 10.704 us | 0.1615 us | 0.1432 us | 10.681 us |      0.0763 |           - |           - |               184 B |
|        ParseRawFloat_ParseContext |        10080 |           ? | 11.675 us | 0.0742 us | 0.0694 us | 11.671 us |           - |           - |           - |                   - |
|   ParseRawDouble_CodedInputStream |        10080 |           ? |  5.625 us | 0.0402 us | 0.0356 us |  5.613 us |      0.0839 |           - |           - |               184 B |
|       ParseRawDouble_ParseContext |        10080 |           ? |  4.083 us | 0.0419 us | 0.0350 us |  4.081 us |           - |           - |           - |                   - |
```

jtattermusch commented:
ParseMessagesBenchmark results:

The original _ParseFromByteArray performance is basically the same as the new __ParseFromReadOnlySequence performance, except that initializing the ParseContext is ~40ns slower than initializing the original CodedInputSteam, so there is a one-time overhead when parsing a single message (one can best see this in the parse EmptyMessage benchmark). This overhead seems completely negligible whenever we start parsing non-trivial messages and such small overhead also can't be noticed in e.g. RPC scenarios where the overhead of an RPC call will be many times larger than this.

The new _ParseFromByteArray scenario still uses CodedInputStream for parsing the messages, so there's extra overhead (once per message parsed) converting the CodedInputStream's internal state into a parseContext (which requires a struct copy). Under normal circumstances this overhead will only be incurred at the root message of the parsing tree and it's unlikely to matter in real-life scenarios.

```
BEFORE

BenchmarkDotNet=v0.11.4, OS=macOS Mojave 10.14.6 (18G3020) [Darwin 18.7.0]
Intel Core i7-7567U CPU 3.50GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT


|                                                         Method | messageCount |         Mean |       Error |      StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|--------------------------------------------------------------- |------------- |-------------:|------------:|------------:|------------:|------------:|------------:|--------------------:|
|                    ManyWrapperFieldsMessage_ParseFromByteArray |            ? |    381.41 ns |   8.6719 ns |  24.3170 ns |      0.8879 |           - |           - |              1864 B |
|                  ManyPrimitiveFieldsMessage_ParseFromByteArray |            ? |    186.74 ns |   2.5298 ns |   2.2426 ns |      0.4880 |           - |           - |              1024 B |
|                                EmptyMessage_ParseFromByteArray |            ? |     37.50 ns |   0.6567 ns |   0.5821 ns |      0.0533 |           - |           - |               112 B |
|   ManyWrapperFieldsMessage_ParseDelimitedMessagesFromByteArray |           10 |  3,405.47 ns |  48.4968 ns |  40.4970 ns |      8.5106 |           - |           - |             17848 B |
| ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromByteArray |           10 |  1,799.61 ns |  22.4263 ns |  20.9775 ns |      4.5033 |           - |           - |              9448 B |
|   ManyWrapperFieldsMessage_ParseDelimitedMessagesFromByteArray |          100 | 33,946.81 ns | 444.3343 ns | 393.8908 ns |     84.6558 |           - |           - |            177688 B |
| ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromByteArray |          100 | 18,146.69 ns | 356.7617 ns | 424.6995 ns |     44.6472 |           - |           - |             93688 B |

AFTER

BenchmarkDotNet=v0.11.4, OS=macOS Mojave 10.14.6 (18G3020) [Darwin 18.7.0]
Intel Core i7-7567U CPU 3.50GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.15 (CoreCLR 4.6.28325.01, CoreFX 4.6.28327.02), 64bit RyuJIT


|                                                                Method | messageCount |         Mean |       Error |      StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------------------------------------------------- |------------- |-------------:|------------:|------------:|------------:|------------:|------------:|--------------------:|
|                           ManyWrapperFieldsMessage_ParseFromByteArray |            ? |    409.88 ns |   7.4969 ns |   6.6458 ns |      0.9341 |           - |           - |              1960 B |
|                    ManyWrapperFieldsMessage_ParseFromReadOnlySequence |            ? |    381.03 ns |   8.7381 ns |   8.1736 ns |      0.8464 |           - |           - |              1776 B |
|                         ManyPrimitiveFieldsMessage_ParseFromByteArray |            ? |    251.87 ns |   4.6516 ns |   4.1235 ns |      0.5336 |           - |           - |              1120 B |
|                  ManyPrimitiveFieldsMessage_ParseFromReadOnlySequence |            ? |    223.52 ns |   4.4897 ns |   6.4390 ns |      0.4461 |           - |           - |               936 B |
|                                       EmptyMessage_ParseFromByteArray |            ? |    106.58 ns |   0.9453 ns |   0.8380 ns |      0.0991 |           - |           - |               208 B |
|                                EmptyMessage_ParseFromReadOnlySequence |            ? |     71.74 ns |   1.0780 ns |   1.0084 ns |      0.0113 |           - |           - |                24 B |
|          ManyWrapperFieldsMessage_ParseDelimitedMessagesFromByteArray |           10 |  3,906.78 ns |  41.3674 ns |  36.6711 ns |      8.5526 |           - |           - |             17944 B |
|   ManyWrapperFieldsMessage_ParseDelimitedMessagesFromReadOnlySequence |           10 |  3,483.34 ns |  39.6573 ns |  37.0955 ns |      8.4648 |           - |           - |             17760 B |
|        ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromByteArray |           10 |  2,235.63 ns |  21.6903 ns |  19.2279 ns |      4.5471 |           - |           - |              9544 B |
| ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromReadOnlySequence |           10 |  1,820.10 ns |  25.1990 ns |  22.3383 ns |      4.4613 |           - |           - |              9360 B |
|          ManyWrapperFieldsMessage_ParseDelimitedMessagesFromByteArray |          100 | 39,119.34 ns | 493.0260 ns | 461.1768 ns |     84.7168 |           - |           - |            177784 B |
|   ManyWrapperFieldsMessage_ParseDelimitedMessagesFromReadOnlySequence |          100 | 36,101.14 ns | 678.6854 ns | 634.8427 ns |     84.6558 |           - |           - |            177600 B |
|        ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromByteArray |          100 | 22,013.79 ns | 278.3219 ns | 260.3425 ns |     44.7083 |           - |           - |             93784 B |
| ManyPrimitiveFieldsMessage_ParseDelimitedMessagesFromReadOnlySequence |          100 | 18,053.55 ns | 299.3745 ns | 280.0351 ns |     44.6167 |           - |           - |             93600 B |
```

jtattermusch commented on csharp/src/Google.Protobuf/ParseContext.cs:
```diff
@@ -0,0 +1,332 @@
+            input.InternalState = state;
+        }
+
+        internal void LoadStateFrom(CodedInputStream input)
```
should be better now.

jtattermusch commented on csharp/src/Google.Protobuf/ParseContext.cs:
```diff
@@ -0,0 +1,332 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void ReadMessage(IMessage message)
+        {
+            // TODO: add a fallback if IMessage does not implement IBufferMessage 
```
Removed (actually moved it elsewhere with extra explanation of what the potential optimization could be.)

jtattermusch commented:
@haberman  this PR is a rewrite of https://github.com/protocolbuffers/protobuf/pull/5888 and should be addressing all the concerns we've had (it's smaller, leads to less duplication and also exposes a new "internal only" InternalMergeFrom() method). I still need to do a bit of polishing but 99% of the stuff is ready - can you please review?

jtattermusch commented on csharp/src/Google.Protobuf/SegmentedBufferHelper.cs:
```diff
@@ -0,0 +1,295 @@
+            {
+                // TODO(jtattermusch): try to initialize the first segment, otherwise the
+                // very first read will result in slowpath (because the first thing to do is to
+                // refill to get the first buffer segment)
```
Improved tests so that his branch is now exercised too. Also addressed the TODO.

JamesNK commented on csharp/src/Google.Protobuf/FieldCodec.cs:
```diff
@@ -583,10 +583,10 @@ internal static FieldCodec<T> GetCodec<T>()
+                return (ValueReader<T?>)value;
             }
 
             internal static T Read<T>(CodedInputStream input, FieldCodec<T> codec)
```
This method isn't used anymore.

JamesNK commented on csharp/src/Google.Protobuf/Collections/MapField.cs:
```diff
@@ -666,6 +690,45 @@ public void MergeFrom(CodedInputStream input)
                 }
 
+                [SecuritySafeCritical]
+                public void InternalMergeFrom(ref ParseContext ctx)
```
The `public void MergeFrom(CodedInputStream input)` on MessageAdapter has no test coverage.

jtattermusch replied:
refactored a bit to deduplicate the code and added a test.

jtattermusch commented on csharp/src/Google.Protobuf/ParsingPrimitives.cs:
```diff
@@ -0,0 +1,726 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ByteString ReadBytes(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state)
+        {
+            int length = ParsingPrimitives.ParseLength(ref buffer, ref state);
```
should be fixed now.

jtattermusch commented on csharp/src/Google.Protobuf/ParsingPrimitivesMessages.cs:
```diff
@@ -0,0 +1,227 @@
+                // Regenerating the code from .proto files will remove this overhead because it will
+                // generate the MergeFrom_Internal method we need.
+
+                if (ctx.state.codedInputStream == null)
```
This is now tested by LegacyGeneratedCodeTest.

jtattermusch commented:
@jskeet FYI

jskeet commented:
@jtattermusch: How does this interact with #7382? Do we need to be careful about the order in which they're merged?

jtattermusch commented:
@jskeet  I don't think the merge conflict is going to be too bad (we're mostly touching different parts of code). I am happy with yielding the right of way and letting you merge first.

jskeet commented:
@jtattermusch: Much appreciated. I'll try to get it in ASAP (without rushing, of course.)

JamesNK commented on csharp/src/Google.Protobuf/SegmentedBufferHelper.cs:
```diff
@@ -0,0 +1,297 @@
+        /// Returns whether or not all the data before the limit has been read.
+        /// </summary>
+        /// <returns></returns>
+        public static bool IsReachedLimit(ref ParserInternalState state)
```
Did you consider making static helper methods like this an instance method on ParserInternalState?

I wonder whether there are any performance benefits of that over passing state by ref to static helper methods.

jtattermusch replied:
I didn't try that but it seems unlikely that it would have any effect on performance.
I think under the hood invoking an instance method on a struct is practically the same thing as invoking a static method that takes a ref (because inside an instance method of a struct "this" is basically a ref to the struct instance).

Nevertheless we could experiment as a followup (but this is just one of the things in the long list of what we could try).

JamesNK commented on csharp/src/Google.Protobuf/SegmentedBufferHelper.cs:
```diff
@@ -0,0 +1,297 @@
+
+        public int? TotalLength => totalLength;
+
+        public CodedInputStream CodedInputStream => codedInputStream;
```
Not used

jtattermusch replied:
Done by removing codedInputStream field from ParserInternalState (which also reduced the size of the struct that needs to be copied in legacy scenarios).

JamesNK commented on csharp/src/Google.Protobuf/SegmentedBufferHelper.cs:
```diff
@@ -0,0 +1,297 @@
+            state.bufferSize = 0;
+            while (readOnlySequenceEnumerator.MoveNext())
+            {
+                
```
nit: extra line

jtattermusch commented:
@gerben-s  sorry, the commentary on benchmark results was buried in the comment thread:
https://github.com/protocolbuffers/protobuf/pull/7351#issuecomment-614558307
https://github.com/protocolbuffers/protobuf/pull/7351#issuecomment-614565505

In short, the results seem to be pretty acceptable, especially because for C# we haven't really been too stringent about C# performance measurements until now (many of the benchmarks have only been added very recently) and we've demonstrated that being able to parse from ReadOnlySequence means a big improvement in terms of avoiding buffer copies and reducing GC pressure (so slight regression for some benchmarks would be trade off worth taking).

JamesNK commented:
Note that one benefit **not** shown by the benchmark results is the cost of creating the array when using `CodedInputStream`. The input stream benchmarks are reading from an already initialized array. In the real world, e.g. gRPC, a framework will need to allocate an array and copy data to it before reading it with `CodedInputStream`.

Reading from a sequence (the newly added feature) allows parsing a third-party buffer of data, e.g. the request stream. That avoids overhead of allocations, copying and garbage collection.

jtattermusch commented:
@gerben-s Thanks for the review! James has already reviewed the code from the C# perspective. I'll see if @jskeet  has any more comments and if not, I'll consider this ready to merge (upon green tests of course).

jtattermusch commented:
@jskeet  do you have any more comments beyond what's been said above? If you want to take some more time to review this, feel free to - there's no rush, just let me know.

jtattermusch commented:
@jskeet  please let me know if you're planning to review this or not.

jtattermusch commented:
Test failures seem completely unrelated to C#.

jtattermusch commented:
Thanks everyone for the reviews and for all the work to get things to this state.

jtattermusch commented:
FTR @jskeet  has discovered a problem with this PR.
Even though our TestProtos project is set to `<LangVersion>3</LangVersion>` to check if `ref struct ParseContext` is fine to use in the generated code, the compilation still seems to fail if we actually use an old compiler:

```
 it doesn't build with a real C# 5 compiler. <LangVersion>5</LangVersion> doesn't pick it up, but fortunately the C# 5 compiler (the last one before Roslyn) is still part of Windows, so I could test it reasonably simply:

> c:\Windows\Microsoft.NET\Framework\v4.0.30319\csc /r:..\Google.Protobuf\bin\debug\net45\google.protobuf.dll *.cs /main:Google.Protobuf.Examples.AddressBook.Program

Errors: 
Addressbook.cs(448,14): error CS0619: 'Google.Protobuf.ParseContext' is obsolete: 'Types with embedded references are not supported in this version of your compiler.'
Addressbook.cs(263,10): error CS0619: 'Google.Protobuf.ParseContext' is obsolete: 'Types with embedded references are not supported in this version of your compiler.'
Addressbook.cs(586,10): error CS0619: 'Google.Protobuf.ParseContext' is obsolete: 'Types with embedded references are not supported in this version of your compiler.'

(The "obsolete" message is unfortunate, but it's an artifact of the way Span/ParseContext and other ref struct are handled by the compiler. That's nothing we've done wrong.)
```

We have some ideas for how to fix this (it will require an `#if`) but we're still not sure it it's going to work out (or we will need to revert).
